File: solutions/code-carl/ch08_greedy/T3_LC0053_MaxSubArray.java
/**
 * Practice template for:
 * - LeetCode 53. Maximum Subarray
 *   https://leetcode.cn/problems/maximum-subarray/
 */
public class T3_LC0053_MaxSubArray {
    public int maxSubArray(int[] nums) {
        int currSum = 0;
        int maxSum = Integer.MIN_VALUE;
        for (int i = 0; i < nums.length; i++) {
            currSum += nums[i];
            maxSum = Math.max(maxSum, currSum);
            if (currSum < 0) {
                currSum = 0;
            }
        }
        return maxSum;
    }
}

File: solutions/python/code-carl/ch08_greedy/t3_lc0053_max_sub_array.py
"""LeetCode 53. Maximum Subarray
https://leetcode.cn/problems/maximum-subarray/
"""

from typing import List

import sys


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        curr_sum: int = 0
        max_sum: int = -sys.maxsize - 1

        for num in nums:
            curr_sum += num
            max_sum = max(curr_sum, max_sum)
            curr_sum = max(curr_sum, 0)

        return max_sum


if __name__ == "__main__":
    solution = Solution()
    assert solution.maxSubArray([-2, 1, -3, 4, -1, 2, 1, -5, 4]) == 6
    assert solution.maxSubArray([1]) == 1
    assert solution.maxSubArray([5, 4, -1, 7, 8]) == 23
    print("All tests passed.")
